DBZ-6835 Documentation for Timezone SMT (#4819)
debezium/debezium: documentation/modules/ROOT/pages/transformations/timezone-converter.adoc

:page-aliases: configuration/timezone-converter.adoc
[id="timezone-converter"]
= Timezone Converter

:toc:
:toc-placement: macro
:linkattrs:
:icons: font
:source-highlighter: highlight.js

toc::[]

[[timezone-converter-introduction]]

> Effortlessly synchronizing timestamp data from diverse sources into a consistent timezone representation holds significance for maintaining data consistency and precision within data processing pipelines and applications. The `Timezone Converter` SMT is designed to seamlessly transform the timestamp values of fields from one timezone to another. It is specifically helpful when you require to ensure that timestamps in your event data are in a consistent timezone representation.

Reviewer: ```suggestion
Effortlessly synchronizing timestamp data from diverse sources into a consistent timezone representation holds significance for maintaining data consistency and precision within data processing pipelines and applications. The `Timezone Converter` SMT is designed to seamlessly transform the timestamp values of fields from one timezone to another. It is specifically helpful when you need to ensure that timestamps in your event data are consistent in timezone representation.
```
While I know we have yet to be consistent about this, we generally strive to have a single sentence per line as it makes edits and future changes easier to merge across different commits.

Author: Thank you for the suggestion, @Naros. I will take it into consideration and make the necessary modifications to the style in the next commit. I wasn't very familiar with the documentation style, so please forgive my novice writing :)

> The SMT converts the value of the specified field to the target timezone using `converted.timezone` configuration option. The target timezone can be specified as a geographic timezone e.g. `America/New_York` or as a UTC offset e.g. `+02:00`. It is assumed that the fields of the record are in UTC. Along with specified timezone, the SMT also provides configuration options to include or exclude specific fields from timezone conversion using the `include.fields` and `exclude.fields` configuration options.

Reviewer: ```suggestion
The SMT converts the value of the specified field to the target timezone using `converted.timezone` configuration option. The target timezone can be specified as a geographic timezone, e.g. `America/New_York`, or as a UTC offset, e.g. `+02:00`. It is assumed that the fields of the record are in UTC. Along with the specified timezone, the SMT also provides configuration options to include or exclude specific fields from timezone conversion using the `include.fields` and `exclude.fields` configuration options.
```

The SMT supports all Debezium and Kafka Connect temporal and non-temporal types.

[NOTE]
====

> In order to comply the conversions with the daylight saving time, you must specify a geographic timezone in the `converted.timezone` configuration option. If you specify a UTC offset, the conversion does not comply with the daylight saving time, it simply implies a fixed offset from UTC. Providing a fixed UTC offset is useful when you want to convert the timestamp fields to a specific timezone that does not follow the daylight saving time.

Reviewer: ```suggestion
To comply with daylight savings time, you must specify a geographic timezone in the `converted.timezone` configuration option. If you specify a UTC offset, the conversion will apply a fixed offset from UTC that isn't accurate for regions that observe daylight savings time. Providing a fixed UTC offset is useful when converting the timestamp fields to a specific timezone that does not follow the daylight saving time.
```

====

[NOTE]
====
The `include.fields` and `exclude.fields` configuration options are mutually exclusive. You must specify only one of the options.
====

[[timezone-converter-usage]]

[[basic-example-timezone-converter]]
== Example: Basic configuration

> To convert the time based fields in an event record to a target timezone, configure the `TimezoneConverter` SMT in the Kafka Connect configuration for a connector.
> For example, to convert the all timestamp fields in an event record from UTC to the `Pacific/Easter` timezone:

Reviewer: ```suggestion
Configure the `TimezoneConverter` SMT in the Kafka Connect configuration for a connector to convert the time-based fields in an event record to a target timezone.
For example, to convert all timestamp fields in an event record from UTC to the `Pacific/Easter` timezone:
```

[source]
----
transforms=convertTimezone
transforms.convertTimezone.type=io.debezium.transforms.TimezoneConverter
transforms.convertTimezone.converted.timezone=Pacific/Easter
----

The following example shows how to convert the all timestamp fields in an event record:

.Effect of applying the `TimezoneConverter` SMT
====
Event record before it is processed by the `TimezoneConverter` transformation::

Value before the SMT processes the event record:::

> +

Reviewer 2: The rendering of the definition list within the example block results in confusing output. I know that this appears elsewhere in the SMT docs, but I strongly disagree with the formatting. The two DL entries are redundant, and they are not clearly separated in the output.  I think that the alternative approach in my comment works better (4th-level heading> intro sentence> example block title>example block).

            "xmin": null
        },
        "op": "c",
        "ts_ms": 1559033904863
    }
----

Event record after it is processed by the `TimezoneConverter` transformation::

Value after the SMT converts the timestamp field to the target timezone:::
+
[source, json]
----
{
        "before": null,
        "after": {
            "id": 1,
            "first_name": "Anne",
            "last_name": "Kretchmar",
            "email": "annek@noanswer.org",
            "created_at": "2011-01-11T11:40:30.123456789-05:00"
        },
        "source": {
            "version": "2.4.0.Aplha2",
            "connector": "postgresql",
            "name": "PostgreSQL_server",
            "ts_ms": 1559033904863,
            "snapshot": true,
            "db": "postgres",
            "sequence": "[\"24023119\",\"24023128\"]",
            "schema": "public",
            "table": "customers",
            "txId": 555,
            "lsn": 24023128,
            "xmin": null,
            "id": 100
        },
        "op": "c",
        "ts_ms": 1559033904863,
        "event_timestamp": 1626102708861
    }
----
====

[[advanced-example-timezone-converter]]
== Example: Advanced configuration

> For example, to convert the `created_at`, and `updated_at` timestamp fields in an event record from UTC to a fixed offset of `+05:30`, configure the `include.fields` configuration option as follows:

Reviewer 2: ```suggestion
Instead of converting all  timestamp fields in an event record, you can configure the SMT to convert specific fields only. 
The following example shows how you might use the `include.fields` option in the SMT configuration to add convert only the `created_at`, and `updated_at` timestamp fields in an event record.
Instead of specifying a geographic timezone designator,  this configuration uses a fixed offset to convert the  time from UTC to `+05:30`.

to convert all timestamp fields in an event record from UTC to the `Pacific/Easter` timezone, add the following lines to your connector configuration:
```

Author: > to convert all timestamp fields in an event record from UTC to the `Pacific/Easter` timezone, add the following lines to your connector configuration:

If my understanding is accurate, this portion appears to be unrelated to the preceding example and seems to have been included in the suggested revisions by mistake.

Reviewer 2: Yes, @ani-sha, you are correct. Seems that I mistakenly pasted this duplicate fragment from part of an earlier sentence.  Apologies for causing any confusion!

[source]
----
transforms=convertTimezone
transforms.convertTimezone.type=io.debezium.transforms.TimezoneConverter
transforms.convertTimezone.converted.timezone=+05:30
transforms.convertTimezone.include.fields=source:customers:created_at,customers:updated_at
----

> In the same the way, you can exclude specific timestamp fields from timezone conversion. For example, to exclude the `updated_at` timestamp field in an event record from timezone conversion, configure the `exclude.fields` configuration option as follows:

Reviewer: ```suggestion
In the same way, you can exclude specific timestamp fields from timezone conversion. For example, to exclude the `updated_at` timestamp field in an event record from timezone conversion, configure the `exclude.fields` configuration option as follows:
```

[source]
----
transforms=convertTimezone
transforms.convertTimezone.type=io.debezium.transforms.TimezoneConverter
transforms.convertTimezone.converted.timezone=+05:30
transforms.convertTimezone.exclude.fields=source:customers:updated_at
----

[[timezone-converter-configuration-options]]
== Configuration options

> The following table lists the configuration options that you can use with the `TimezoneConverter` SMT.

Reviewer: ```suggestion
The following table lists the configuration options for the `TimezoneConverter` SMT.
```

.TimezoneConverter SMT configuration options
[cols="14%a,40%a,10%a,10%a"]
|===
|Property
|Description
|Type
|Importance

|[[timezone-converter-converted-timezone]]<<timezone-converter-converted-timezone, `converted.timezone`>>

> |A string that specifies the target timezone to which the timestamp fields should be converted. The target timezone can be specified as a geographic timezone e.g. `America/New_York` or as a UTC offset e.g. `+02:00`.

Reviewer: ```suggestion
|A string that specifies the target timezone to which the timestamp fields should be converted. The target timezone can be specified as a geographic timezone, e.g. `America/New_York`, or as a UTC offset, e.g. `+02:00`.
```

|string
|high
|[[timezone-converter-include-fields]]<<timezone-converter-include-fields, `include.fields`>>

> |A comma-separated list of rules that specify which fields should be included for timezone conversion, using one of the following formats:

Reviewer: ```suggestion
|A comma-separated list of rules that specify which fields should be included for timezone conversion with one of the following formats:
```

>     `source:<tablename>` :: Matches Debezium change events with a source information block having the specified table name. All time-based fields in the matched table will be converted.

Reviewer 2: ```suggestion
    `source:<tablename>` :: Matches {prodname} change events with source information blocks that have the specified table name. 
The SMT converts all time-based fields in the matched table.
```

>     `source:<tablename>:<fieldname>` :: Matches Debezium change events with a source information block having the specified table name. Only the specified field name will be converted.

Reviewer 2: ```suggestion
    `source:<tablename>:<fieldname>` :: Matches {prodname} change events with source information blocks that have the specified table name. 
The SMT converts only fields in the specified table that have the specified field name.
```

>     `topic:<topicname>` :: Matches events from the specified topic name, converting all time-based fields.

Reviewer 2: ```suggestion
    `topic:<topicname>` :: Matches events from the specified topic name, converting all time-based fields in the event record.
```

>     `topic:<topicname>:<fieldname>` :: Matches events from the specified topic name, converting only the specified field name.

Reviewer 2: ```suggestion
    `topic:<topicname>:<fieldname>` :: Matches events from the specified topic name, and converts values for the specified fields only.
```

>     `<matchname>:<fieldname>` :: Applies a heuristic matching algorithm to match against the source information block table name if present; otherwise, matches against the topic name. Conversion is applied only to the specified field name.

Reviewer 2: ```suggestion
    `<matchname>:<fieldname>` :: Applies a heuristic matching algorithm to match against the table name of the source information block, if present; otherwise, matches against the topic name. 
The SMT converts values for the specified field name only.
```

|list
|medium
|[[timezone-converter-exclude-fields]]<<timezone-converter-exclude-fields, `exclude.fields`>>

> |A comma-separated list of rules that specify what fields should be excluded from timezone conversion, using one of the following formats:

Reviewer: ```suggestion
|A comma-separated list of rules that specify what fields should be excluded from timezone conversion with one of the following formats:
```

>     `source:<tablename>` :: Matches Debezium change events with a source information block having the specified table name. All time-based fields in the matched table will be excluded from conversion.

Reviewer 2: ```suggestion
    `source:<tablename>` :: Matches {prodname} change events with source information blocks that have the specified table name.
 The SMT excludes all time-based fields in the matched table from conversion.
```

>     `source:<tablename>:<fieldname>` :: Matches Debezium change events with a source information block having the specified table name. Only the specified field name will be excluded from conversion.

Reviewer 2: ```suggestion
    `source:<tablename>:<fieldname>` :: Matches {prodname} change events with source information blocks that have the specified table name. 
The SMT excludes from conversion fields in the specified table that match the specified field name.
```

>     `topic:<topicname>` :: Matches events from the specified topic name, excluding all time-based fields.

Reviewer 2: ```suggestion
    `topic:<topicname>` :: Matches events from the specified topic name, and excludes from conversion all time-based fields in the topic.
```

>     `topic:<topicname>:<fieldname>` :: Matches events from the specified topic name, excluding only the specified field name.

Reviewer 2: ```suggestion
    `topic:<topicname>:<fieldname>` :: Matches events from the specified topic name, and excludes from conversion any fields in the topic that have the specified name.
```

>     `<matchname>:<fieldname>` :: Applies a heuristic matching algorithm to match against the source information block table name if present; otherwise, matches against the topic name. Excludes conversion only for the specified field name.

Reviewer 2: ```suggestion
    `<matchname>:<fieldname>` :: Applies a heuristic matching algorithm to match against the against the table name of the source information block, if present; otherwise, matches against the topic name. 
The SMT excludes from conversion only fields that have the specified name.
```

:page-aliases: configuration/timezone-converter.adoc
[id="timezone-converter"]
= Timezone Converter

:toc:
:toc-placement: macro
:linkattrs:
:icons: font
:source-highlighter: highlight.js

toc::[]

[[timezone-converter-introduction]]

> Effortlessly synchronizing timestamp data from diverse sources into a consistent timezone representation holds significance for maintaining data consistency and precision within data processing pipelines and applications.

Reviewer 2: ```suggestion
When {prodname} emits event records, the timezone values for timestamp fields in the record can vary depending on the type and configuration of the data source.
As a result, events that originate from different sources can contain timestamp fields that contain inconsistent timezone values. 
Such inconsistent values can be problematic for downstream applications, which are designed to consume timestamp data that conforms to a common format.  
```

> The `Timezone Converter` SMT is designed to seamlessly transform the timestamp values of fields from one timezone to another.

Reviewer 2: ```suggestion
To maintain data consistency and precision within data processing pipelines and applications, you can use the `Timezone Converter` SMT to convert the time zone values of timestamp fields so that they represent data consistently.

```

> To maintain data consistency and precision within data processing pipelines and applications, you can use the `Timezone Converter` SMT to convert the time zone values of timestamp fields so that they represent data consistently.
> It is specifically helpful when you need to ensure that timestamps in your event data are consistent in timezone representation.

Reviewer 2: ```suggestion
To maintain data consistency and precision within data processing pipelines and applications, you can use the `Timezone Converter` SMT to ensure that event records use a consistent timezone to represent timestamp data.
```

Reviewer 2: These two sentences are partially redundant. Does this successfully combine the thoughts from both of them?

> The SMT converts the value of the specified field to the target timezone using `converted.timezone` configuration option.

Reviewer 2: ```suggestion
The SMT converts the value of the specified field to the target timezone by using the `converted.timezone` configuration option.
```

> The target timezone can be specified as a geographic timezone, for example,  `America/New_York`, or as a UTC offset, such as `+02:00`.

Reviewer 2: ```suggestion
You can specify the target timezone as a geographic timezone, for example,  `America/New_York`, or as a UTC offset, such as `+02:00`.
```

[NOTE]
====
To comply with daylight savings time, you must specify a geographic timezone in the `converted.timezone` configuration option.
If you specify a UTC offset, the transform applies a fixed offset from UTC that isn't accurate for regions that observe daylight savings time.
Providing a fixed UTC offset is useful when converting timestamp fields to a specific timezone that does not observe daylight saving time.
====

[NOTE]
====
The `include.fields` and `exclude.fields` configuration options are mutually exclusive. You must specify only one of the options.
====

[[timezone-converter-usage]]

[[basic-example-timezone-converter]]
== Example: Basic configuration

Configure the `TimezoneConverter` SMT in the Kafka Connect configuration for a connector to convert the time-based fields in an event record to a target timezone.

For example, to convert all timestamp fields in an event record from UTC to the `Pacific/Easter` timezone, add the following lines to your connector configuration:

[source]
----
transforms=convertTimezone
transforms.convertTimezone.type=io.debezium.transforms.TimezoneConverter
transforms.convertTimezone.converted.timezone=Pacific/Easter
----

> === Effect of applying the `TimezoneConverter` SMT
> The following examples show the timestamp fields in an event record before and after applying the `TimezoneConverter` transformation.

Reviewer 2: ```suggestion
=== Effect of applying the `TimezoneConverter` SMT

The following examples show the timestamp fields in an event record before and after applying the `TimezoneConverter` transformation.
```

I think that it's best to insert a blank line between headings and the first sentence that follows. It's inconsistent, but I've seen some instances where omitting an empty line results in the first sentence being dropped when content is converted for downstream use.

Author: @roldanbob This creates the table of contents to be in this structure as I see in the preview:

Table of Contents
- [Example: Basic configuration](https://github.com/debezium/debezium/pull/4819#basic-example-timezone-converter)
    - [Effect of applying the TimezoneConverter SMT](https://github.com/debezium/debezium/pull/4819#_effect_of_applying_the_timezoneconverter_smt)
- [Example: Advanced configuration](https://github.com/debezium/debezium/pull/4819#advanced-example-timezone-converter)
- [Configuration options](https://github.com/debezium/debezium/pull/4819#timezone-converter-configuration-options)

Author: Also inside the `Effect of applying the TimezoneConverter SMT` block the headings are being shown as 
```
Example 1. Value of created_at field in an event record before processing by the TimezoneConverter transformation
``` 
and 
```
Example 2. Event record value after processing by the TimezoneConverter transformation
``` 
which I believe are part of the same example and shouldn't be depicted as Example 1 & 2.  it.

Reviewer 2: You seemed to be saying that adding the blank line results in the TOC entry, but I don't think that's what you meant? In any case, I don't think that adding the subordinate TOC entry is a problem. 
WRT your second comment,  I suggest that we retain the _before_ and _after_ examples as separate example blocks. They are both components in demonstrating the concept of the parent heading, i.e., Effect of applying the SMT ... 

But my labeling change needs improvement. The two examples should have parallel titles. To achieve that, the first should be _"Event record value before processing by ..."_ . I've edited my original suggestion to reflect that.

> .Value of `created_at` field in an event record before processing by the `TimezoneConverter` transformation

Reviewer 2: ```suggestion
.Event record value before processing by the `TimezoneConverter` transformation
```

====
The SMT converts the original UTC value of the `created_at` field to the time in the target `Pacific/Easter` timezone that is specified in the xref:basic-example-timezone-converter[Basic configuration] example.
The SMT also adds an `event_timestamp` field.

[source, json]
----
{
        "before": null,
        "after": {
            "id": 1,
            "first_name": "Anne",
            "last_name": "Kretchmar",
            "email": "annek@noanswer.org",
            "created_at": "2011-01-11T11:40:30.123456789-05:00"
        },
        "source": {
            "version": "2.4.0.Aplha2",
            "connector": "postgresql",
            "name": "PostgreSQL_server",
            "ts_ms": 1559033904863,
            "snapshot": true,
            "db": "postgres",
            "sequence": "[\"24023119\",\"24023128\"]",
            "schema": "public",
            "table": "customers",
            "txId": 555,
            "lsn": 24023128,
            "xmin": null,
            "id": 100
        },
        "op": "c",
        "ts_ms": 1559033904863,
        "event_timestamp": 1626102708861
    }
----
====

[[advanced-example-timezone-converter]]
== Example: Advanced configuration

> Instead of converting all  timestamp fields in an event record, you can configure the SMT to convert specific fields only.

Reviewer 2: ```suggestion
Instead of converting all timestamp fields in an event record, you can configure the SMT to convert specific fields only.
```

> The following example shows how you might use the `include.fields` option in the SMT configuration to add convert only the `created_at`, and `updated_at` timestamp fields in an event record.

Reviewer 2: ```suggestion
The following example shows how you might use the `include.fields` option in the SMT configuration to convert only the `created_at`, and `updated_at` timestamp fields in an event record.
```

> Instead of specifying a geographic timezone designator,  this configuration uses a fixed offset to convert the  time from UTC to `+05:30`.

Reviewer 2: ```suggestion
Instead of specifying a geographic timezone designator, this configuration uses a fixed offset to convert the time from UTC to `+05:30`.
```

> Instead of specifying a geographic timezone designator, this configuration uses a fixed offset to convert the  time from UTC to `+05:30`.

Reviewer 2: ```suggestion
The following configuration uses a fixed offset, rather than a geographic timezone designator,  to convert the time from UTC to `+05:30`.
```

[source]
----
transforms=convertTimezone
transforms.convertTimezone.type=io.debezium.transforms.TimezoneConverter
transforms.convertTimezone.converted.timezone=+05:30
transforms.convertTimezone.include.fields=source:customers:created_at,customers:updated_at
----

> Similarly, you can exclude specific timestamp fields from timezone conversion.

Reviewer 2: ```suggestion
In some cases, you might want to exclude specific timestamp fields from timezone conversion.
```

> For example, to exclude the `updated_at` timestamp field in an event record from timezone conversion, us the `exclude.fields` configuration option as in the following example:

Reviewer 2: ```suggestion
For example, to exclude the `updated_at` timestamp field in an event record from timezone conversion, use the `exclude.fields` configuration option as in the following example:
```
